File: PRACTICAS/practica1/Civitas_Ruby/lib/casilla.rb
require './tipo_sorpresa'
require './tipo_casilla'
module Civitas
class Casilla
  
  
  @@carcel
 
  
  def initialize (nombre = "sin_nombre", carcel = 0, importe = 0, tipo = nil, titulo_propiedad = nil, mazo = nil, sorpresa = nil)
    
    @nombre = nombre  
    @@carcel = carcel
    @importe = importe
    @tipo = tipo
    @titulo_propiedad = titulo_propiedad
    @mazo = mazo
    @sorpresa = sorpresa
    
  end
  
  def self.casilla(titulo_propiedad )
   
    self.new("sin_nombre", 0, 0, nil, titulo_propiedad, nil, nil)
    
  end
  
  def self.casilla2(cantidad,nombre)
    
    self.new(nombre, 0, cantidad, nil, nil, nil, nil)
    
  end
  
  def self.casilla3(num_casilla_carcel, nombre)
    
    self.new(nombre, num_casilla_carcel, 0, nil, nil, nil, nil)
    
  end
  
  def self.casilla4(mazo, nombre)
   
    self.new(nombre, 0, 0, nil, nil, mazo, nil)
    
  end
  
  
  attr_reader :nombre          #es el getNombre()
  attr_reader :titulo_propiedad
  
  def informe(iactual,todos)
    
    Diario.instance.ocurre_evento("El jugador " + todos[iactual].nombre + "ha caido en la casilla")
    to_string
    
  end
  private :informe

  
  def jugador_correcto(iactual,todos)
    
    
    return (iactual > 0 && iactual < todos.size)
    
  end
  
  public :jugador_correcto
  
  def recibe_jugador(iactual,todos)
    
    
    
  end
  
  def recibe_jugador_calle(iactual,todos)
    
  end
  
  private :recibe_jugador_calle
  
  def recibe_jugador_impuesto(iactual,todos)
    
    if (jugador_correcto(iactual, todos))    
      informe(iactual, todos)
      todos[iactual].paga_impuesto(@importe)
            
        
    end
  end
  
  private :recibe_jugador_impuesto
  
  def recibe_jugador_juez(iactual,todos)
    
    if (jugador_correcto(iactual, todos))
      informe(iactual, todos)
      todos[iactual].encarcelar(@carcel)
        
    
    end
  end
  
  private :recibe_jugador_juez
  
  
  def recibe_jugador_sorpresa(iactual,todos)
    
  end
  
  private :recibe_jugador_sorpresa
  
  def to_string()
    puts "Nombre: #{@nombre} y su importe es de: #{@importe} y es de tipo: #{@tipo} y la posicion en la carcel es: #{@@carcel}"
  end
  
  public :to_string
  
  
  
end
end

File: PRACTICAS/CAMBIAR_RUBY/COMPROBACIONES JAVA/src/civitas_p1/Casilla.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor. eeeeeeeeeeeeeee
 */
package civitas_p1;

import java.util.ArrayList;

/**
 *
 * @author angelsolano
 */
public class Casilla {
    
    private String nombre;
    private static int carcel;
    private float importe;
    private TipoCasilla tipo;
    private TituloPropiedad tituloPropiedad;
    private MazoSorpresas mazo;
    private Sorpresa sorpresa;
    
   
    Casilla(String nombre){
        
        init();
        this.nombre = nombre;
        tipo = TipoCasilla.DESCANSO;
        
    }
    
    Casilla(TituloPropiedad titulo){
        init();
        tituloPropiedad = titulo;
        nombre = titulo.getNombre();
        tipo = TipoCasilla.CALLE;       
        importe = titulo.getPrecioCompra();
    }
    
    Casilla(float cantidad, String nombre){
        init();
        this.nombre = nombre;
        importe = cantidad;
        tipo = TipoCasilla.IMPUESTO;
    }
    
    Casilla(int numCasillaCarcel, String nombre){
        init();
        this.nombre = nombre;
        carcel = numCasillaCarcel;
        tipo = TipoCasilla.JUEZ;
    }
    
    Casilla(MazoSorpresas mazo, String nombre){
        init();
        this.nombre = nombre;
        this.mazo = mazo;
        tipo = TipoCasilla.SORPRESA;
        
    }
    
    private void init(){
        nombre = "UNDEFINED";
        carcel = 8;
        importe = (float) 0.0;
        tituloPropiedad = null;
        mazo = null;
        sorpresa = null;
        
        
    }
    
    public String getNombre(){
        return nombre;
    }
    
    private void informe(int iactual, ArrayList<Jugador> todos){
        
        Diario.instance.ocurreEvento("El jugador " + todos.get(iactual).getNombre() + " ha caido en la casilla: " + toString());
        
    }
    
    void recibeJugador(int iactual, ArrayList<Jugador> todos){

        switch (tipo) {
            case CALLE:
                System.out.println("\nRECIBE JUGADOR CALLE\n");
                recibeJugador_calle(iactual, todos);
                
                break;
            case IMPUESTO:
                recibeJugador_impuesto(iactual, todos);
                System.out.println("\nRECIBE JUGADOR IMPUESTO\n");
                break;
            case JUEZ:
                recibeJugador_juez(iactual, todos);
                System.out.println("\nRECIBE JUGADOR JUEZ\n");
                break;
            case SORPRESA:
                recibeJugador_sorpresa(iactual, todos);
                System.out.println("\nRECIBE JUGADOR SORPRESA\n");
                break;
            default:
                informe(iactual, todos);
                System.out.println("\nHACE EL INFORME\n");
                break;
        }

        
    }
    
    private void recibeJugador_calle(int iactual, ArrayList<Jugador> todos){
        
  
        
        if(jugadorCorrecto(iactual, todos)){
            
            informe(iactual, todos);
            Jugador jugador = new Jugador(todos.get(iactual));
         
            
            if(!tituloPropiedad.tienePropietario()){
                
                jugador.puedeComprarCasilla();
            }else{
                
                tituloPropiedad.tramitarAlquiler(jugador);
                      
                
            }
           
        }
       
    }
    
    
    
    private void recibeJugador_impuesto(int iactual, ArrayList<Jugador> todos){
        if (jugadorCorrecto(iactual, todos)){
            informe(iactual, todos);
            todos.get(iactual).pagaImpuesto(importe);
        }
    }
    
    private void recibeJugador_juez(int iactual, ArrayList<Jugador> todos){
        if (jugadorCorrecto(iactual, todos)){
            informe(iactual, todos);
            todos.get(iactual).encarcelar(carcel);
        }              
    }
    
    private void recibeJugador_sorpresa(int iactual, ArrayList<Jugador> todos){
    
        if(jugadorCorrecto(iactual,todos)){

            sorpresa = mazo.siguiente();
            informe(iactual,todos);
            sorpresa.aplicarAJugador(iactual, todos);

        }
        
    }
    
    public boolean jugadorCorrecto(int iactual, ArrayList<Jugador> todos){
        return iactual >= 0 && iactual < todos.size();
    }
    
    @Override
    public String toString(){   
        return "Casilla llamada " + nombre + ", la carcel está en la posición " + carcel + ", el importe es de " + importe + " y es de tipo " + tipo + ".";
    }
    
    TituloPropiedad getTituloPropiedad(){
        return tituloPropiedad;
    }
    
    
    
}
